ruma-appservice-api/src/event/push_events/v1.rs

```diff
@@ -36,6 +44,84 @@ impl<'a> Request<'a> {
+    #[serde(rename = "type")]
+    pub ev_type: String,
+
+    /// If `state_key` is present the event will be deserialized as a state event.
+    pub state_key: Option<IgnoredAny>,
+
+    /// If no `state_key` is found but an `event_id` is present the event
+    /// will be deserialized as a message event.
+    pub event_id: Option<IgnoredAny>,
+
+    /// If no `event_id` or `state_key` are found but a `room_id` is present
+    /// the event will be deserialized as a ephemeral event.
+    pub room_id: Option<RoomId>,
+
+    /// If this `UnsignedData` contains a `redacted_because` key the event is
+    /// immediately deserialized as a redacted event.
+    pub unsigned: Option<UnsignedDeHelper>,
+}
+
+// all observed room events go into "Joined Room" since appservice transactions
+// aren't bound to a specific user
+impl TryInto<sync_events::Response> for IncomingRequest {
+    type Error = serde_json::Error;
+
+    fn try_into(self) -> Result<sync_events::Response, Self::Error> {
+        let mut response = sync_events::Response::new(self.txn_id.clone());
+
+        for raw_event in self.events {
+            let helper: EventDeHelper = serde_json::from_str(raw_event.json().get())?;
+            let event_json = Raw::into_json(raw_event);
+
+            if let Some(room_id) = helper.room_id {
+                let join =
+                    response.rooms.join.entry(room_id).or_insert_with(sync_events::JoinedRoom::new);
+
+                if helper.state_key.is_some() {
+                    match helper.unsigned {
+                        Some(unsigned) if unsigned.redacted_because.is_some() => join
+                            .timeline
+                            .events
+                            .push(Raw::<AnySyncRoomEvent>::from_json(event_json)),
+
+                        _ => {
+                            join.state.events.push(Raw::<AnySyncStateEvent>::from_json(event_json))
+                        }
+                    }
+                } else if helper.event_id.is_some() {
+                    join.timeline.events.push(Raw::<AnySyncRoomEvent>::from_json(event_json));
+                } else {
+                    join.ephemeral
+                        .events
+                        .push(Raw::<AnySyncEphemeralRoomEvent>::from_json(event_json));
+                }
+            } else {
+                #[allow(clippy::single_match)]
+                match helper.ev_type.as_ref() {
+                    "m.presence" => {
+                        response.presence.events.push(Raw::<PresenceEvent>::from_json(event_json))
+                    }
+                    _ => (),
+                }
+            }
+        }
+
+        Ok(response)
+    }
+}
+
+impl IncomingRequest {
+    /// Creates `IncomingRequest` with the given transaction ID and list of events.
+    pub fn new(txn_id: String, events: Vec<Raw<AnyEvent>>) -> IncomingRequest {
+        IncomingRequest { txn_id, events }
+    }
+}
```

Comment:
Nit: Would be nice if this was added in a separate commit, since it's not directly related to the rest of the changes.